test-grapesy/Test/Sanity/Disconnect.hs

```diff
@@ -0,0 +1,288 @@
+          predicates =
+            replicate (numCalls - 1) predicate ++
+              [ \n -> do
+                  when (n == 100) killServer
+                  return False
+              ]
+      results <-
+        mapConcurrently
+          (   try @Client.ServerDisconnected
+            . Client.withRPC conn def (Proxy @Trivial)
+            . countUntil
+          )
+          predicates
+
+      -- All calls should have failed
+      assertBool "" (null (rights results) && length (lefts results) == numCalls)
+
+      -- New calls should succeed (after reconnection)
+      killRestarted <- takeMVar signalRestart
+      result <- Client.withRPC conn def (Proxy @Trivial) $
+        countUntil (pure . (>= 100))
+      assertBool "" (result == 100)
+
+      -- Do not leave the server process hanging around
+      killRestarted
+
+{-------------------------------------------------------------------------------
+  Client and handler functions
+-------------------------------------------------------------------------------}
+
+-- | Send increasing numbers to the server until it responds with one that
+-- satisfies the given predicate.
+countUntil :: forall rpc.
+  ( Input  rpc ~ Lazy.ByteString
+  , Output rpc ~ Lazy.ByteString
+  , ResponseTrailingMetadata rpc ~ NoMetadata
+  ) => (Word64 -> IO Bool) -> Client.Call rpc -> IO Word64
+countUntil = go 0
+  where
+    go :: Word64 -> (Word64 -> IO Bool) -> Client.Call rpc -> IO Word64
+    go next p call = do
+        sat <- p next
+        if sat then do
+          Binary.sendFinalInput @Word64 call next
+          (final, NoMetadata) <- Binary.recvFinalOutput @Word64 call
+          return final
+        else do
+          Binary.sendNextInput @Word64 call next
+          next' <- Binary.recvNextOutput @Word64 call
+          go (succ next') p call
+
+-- | Echos any input
+echoHandler ::
+  ( Input  rpc ~ Lazy.ByteString
+  , Output rpc ~ Lazy.ByteString
+  , ResponseTrailingMetadata rpc ~ NoMetadata
+  ) => Maybe (IORef Int) -> Server.Call rpc -> IO ()
+echoHandler disconnectCounter call = trackDisconnects disconnectCounter $ do
+    Binary.recvInput @Word64 call >>= \case
+      StreamElem n  -> do
+        Binary.sendNextOutput @Word64 call n
+        echoHandler disconnectCounter call
+      FinalElem n _ -> do
+        Binary.sendFinalOutput @Word64 call (n, NoMetadata)
+      NoMoreElems _ -> do
+        Server.sendTrailers call NoMetadata
+  where
+    trackDisconnects Nothing =
+        id
+    trackDisconnects (Just counter) =
+        handle (
+            \(_e :: Server.ClientDisconnected) ->
+              atomicModifyIORef' counter $ \n -> (n + 1, ())
+          )
+
+{-------------------------------------------------------------------------------
+  Auxiliary
+-------------------------------------------------------------------------------}
+
+foreign import ccall unsafe "kill" c_kill :: CInt -> CInt -> IO ()
+foreign import ccall unsafe "exit" c_exit :: CInt -> IO ()
```

Comment:
`exitWith` translates to `throwIO`, so exception handling happens and the test fails. We need the immediate exit `c_exit`.

Reply:
Ah, I see, fair enough. Then let's add least add a comment here, because exiting the Haskell process this way also doesn't allow the RTS to shutdown cleanly. Of course, that's kind of the point here: this is modelling the server crashing for some reason.